Source/Rules/BrokenRule.cs

```diff
@@ -29,28 +29,28 @@ public BrokenRule(IRule rule, object instance, IRuleContext context)
         /// <summary>
         /// Gets the type of rule that is broken
         /// </summary>
-        public IRule Rule { get; private set; }
+        public IRule Rule { get; }
 
         /// <summary>
         /// Gets the instance used for evaluating the <see cref="IRule"/>
         /// </summary>
-        public object Instance { get; private set; }
+        public object Instance { get; }
 
         /// <summary>
         /// Gets the context in which the rule was broken
         /// </summary>
-        public IRuleContext Context { get; private set; }
+        public IRuleContext Context { get; }
 
         /// <summary>
         /// Gets the <see cref="BrokenRuleReason">reasons</see> why the rule is broken
         /// </summary>
-        public IEnumerable<BrokenRuleReason> Reasons { get { return _reasons; } }
+        public IEnumerable<BrokenRuleReasonInstance> Reasons => _reasons;
```

Comment:
Should we do _reasons.ToArray() ?  Just so the external state isn't exposed.  I don't worry about it too much as if you use reflection to discover it's a list and add to it... well, you deserve all you get.  It's not too much further to get access to a private variable via reflection.

What was my point again?

Reply:
Actually from an immutability perspective, that is a good pattern to stick to.